### PR TITLE
Improved quadrangulated disk mesh cmesh example [6/n]

### PR DESCRIPTION
**Pull request by jmark**
**_Describe your changes here:_**

This is a follow-up to https://github.com/DLR-AMR/t8code/pull/792.

This PR improves the squared disk cmesh example by rearranging the initial cmesh elements. This should lead
to a better distribution of characteristic element sizes of same refinement level.
grafik

Added also a 3D version: cubed sphere
grafik
grafik

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

#### General
- [x] The reviewer executed the new code features at least once and checked the results manually

- [x] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [x] New source/header files are properly added to the Makefiles
- [x] The code is well documented
- [ ] All function declarations, structs/classes and their members have a proper doxygen documentation
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)

#### Tests
- [ ] The code is covered in an existing or new test case using Google Test

#### Github action

- [x] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)
- [ ] All tests pass (in various configurations, this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

#### Scripts and Wiki

- [ ] If a new directory with source-files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [ ] If this PR introduces a new feature, it must be covered in an example/tutorial and a Wiki article.

#### Licence

- [x] The author added a BSD statement to `doc/` (or already has one)

### CODE REVIEW

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx [hunk 1]
```diff
@@ -349,6 +329,71 @@ t8_geometry_cubed_spherical_shell::t8_geom_evaluate (t8_cmesh_t cmesh, t8_gloidx
...
+    }
+
+    t8_geom_linear_interpolation (ref_coords + offset, active_tree_vertices, 3, 3, p);
+
+    /* Compute intersection of line with a plane. */
+    const double R = (p[0] * n[0] + p[1] * n[1] + p[2] * n[2]) / (r[0] * n[0] + r[1] * n[1] + r[2] * n[2]);
+
+    out_coords[offset + 0] = (1.0 - z_ref) * p[0] + z_ref * R * s[0];
+    out_coords[offset + 1] = (1.0 - z_ref) * p[1] + z_ref * R * s[1];
+    out_coords[offset + 2] = (1.0 - z_ref) * p[2] + z_ref * R * s[2];
```
Comment (lukasdreyer) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx:393]: There is probably a t8_vec function which does this in a single call

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx [hunk 2]
```diff
@@ -349,6 +329,71 @@ t8_geometry_cubed_spherical_shell::t8_geom_evaluate (t8_cmesh_t cmesh, t8_gloidx
...
+  }
+
+  n[0] = active_tree_vertices[0 + 0];
+  n[1] = active_tree_vertices[0 + 1];
+  n[2] = active_tree_vertices[0 + 2];
+  t8_vec_normalize (n);
+
+  r[0] = active_tree_vertices[7 * 3 + 0];
+  r[1] = active_tree_vertices[7 * 3 + 1];
+  r[2] = active_tree_vertices[7 * 3 + 2];
```
Comment (lukasdreyer) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx:363]: I would love to have some function `get_3d_vertex(active_tree_vertices, 7, r)` to clear up a lot of code, but also probably out of scope of this PR

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx [hunk 3]
```diff
@@ -52,60 +48,44 @@ t8_geometry_squared_disk::t8_geom_evaluate (t8_cmesh_t cmesh, t8_gloidx_t gtreei
...
+    const double y_ref = ref_coords[offset + 1];
 
     {
       double corr_ref_coords[3];
 
       /* Correction in order to rectify elements near the corners. */
-      corr_ref_coords[0] = tan (0.5 * M_PI * (x - 0.5)) * 0.5 + 0.5;
-      corr_ref_coords[1] = y;
+      corr_ref_coords[0] = tan (0.25 * M_PI * x_ref);
+      corr_ref_coords[1] = y_ref;
```
Comment (lukasdreyer) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx:74]: Do you have a reference, as to why you use this kind of correction?

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx (hunk 1 above)
Comment (jmark) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx:393]: Not quite. But I tried hard to use `t8_vec.h` routines.

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx [hunk 4]
```diff
@@ -25,20 +25,16 @@
...
+t8_geometry_quadrangulated_disk::t8_geom_evaluate (t8_cmesh_t cmesh, t8_gloidx_t gtreeid, const double *ref_coords,
+                                                   const size_t num_coords, double *out_coords) const
 {
-  if (num_coords != 1)
-    SC_ABORT ("Error: Batch computation of geometry not yet supported.");
-
   double n[3]; /* Normal vector. */
   double r[3]; /* Radial vector. */
   double s[3]; /* Radial vector for the corrected coordinates. */
   double p[3]; /* Vector on the plane resp. quad. */
```
[the remark this answers is not in the document]
Comment (jmark) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx:34]: Oh, Lukas ... you gonna be very happy. ;)

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx (hunk 2 above)
Comment (jmark) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx:363]: `t8_vec_copy` for the rescue!

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx (hunk 3 above)
Comment (jmark) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx:74]: No, I don't. Divine insight resp. by tinkering. I suspect there actually no analytical function which does a perfect correction.

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx (hunk 4 above)
[the remark this answers is not in the document]
Comment (jmark) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx:34]: Pretty hard to to do in ascii art. Do I really have to burn work hours to make one?

[the remark this answers is not in the document]
Comment (lukasdreyer) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx:34]: No, as I said above, this is probably not in relation to the benefit

**Review by lukasdreyer — APPROVED** [review]

**Pull request merged**